lib/ProductOpener/Auth.pm

```diff
@@ -0,0 +1,743 @@
+	$get_user_request->header('Content-Type' => 'application/json');
+	$get_user_request->header('Authorization' => $token->{token_type} . ' ' . $token->{access_token});
+	my $get_user_response = LWP::UserAgent::Plugin->new->request($get_user_request);
+	unless ($get_user_response->is_success) {
+		display_error_and_exit($get_user_response->content, 500);
+	}
+
+	my $json_response = $get_user_response->decoded_content(charset => 'UTF-8');
+	my @created_users = decode_json($json_response);
+	return $created_users[0];
+}
+
+sub create_user_in_product_opener ($id_token) {
+	unless ($id_token) {
+		return;
+	}
+
+	my $user_ref = {};
+	$user_ref->{email} = $id_token->{'email'};
+	$user_ref->{userid} = $id_token->{'preferred_username'};
+	$user_ref->{name} = $id_token->{'name'};
+
+	my $user_id = $user_ref->{userid};
+	my $user_file = "$BASE_DIRS{USERS}/" . get_string_id_for_lang("no_language", $user_id) . ".sto";
+	store($user_file, $user_ref);
+
+	# Store email
+	my $emails_ref = retrieve("$BASE_DIRS{USERS}/users_emails.sto");
+	my $email = $user_ref->{email};
+
+	if ((defined $email) and ($email =~ /\@/)) {
+		$emails_ref->{$email} = [$user_id];
+	}
+
+	if (defined $user_ref->{old_email}) {
+		delete $emails_ref->{$user_ref->{old_email}};
+		delete $user_ref->{old_email};
+	}
+
+	store("$BASE_DIRS{USERS}/users_emails.sto", $emails_ref);
+	return $user_id;
+}
+
+=head2 get_token_using_password_credentials($username, $password)
+
+Gets a token for the user.
+
+Method uses the Resource Owner Password Credentials Grant to
+with the given credentials, and pre-configured Client ID,
+and Client Secret.
+
+=head3 Arguments
+
+=head4 Name of the user $usersname
+
+=head4 Password given at sign-in $password
+
+=head3 Return values
+
+Open ID Access token, or undefined if sign-in wasn't successful.
+
+=cut
+
+sub get_token_using_password_credentials ($username, $password) {
+	_ensure_oidc_is_discovered();
+
+	my $token_request = HTTP::Request->new(POST => $oidc_discover_document->{token_endpoint});
+	$token_request->header('Content-Type' => 'application/x-www-form-urlencoded');
+	$token_request->content('grant_type=password&client_id='
+			. uri_escape($oidc_options{client_id})
+			. '&client_secret='
+			. uri_escape($oidc_options{client_secret})
+			. '&username='
+			. uri_escape($username)
+			. '&password='
+			. uri_escape($password)
+			. "&scope=openid%20profile%20offline_access");
+
+	my $token_response = LWP::UserAgent::Plugin->new->request($token_request);
+	unless ($token_response->is_success) {
+		$log->info('bad password - no token returned from IdP', {content => $token_response->content})
+			if $log->is_info();
+		return;
+	}
+
+	my $access_token = decode_json($token_response->content);
+	$log->info('got access token from password credentials', {access_token => $access_token}) if $log->is_info();
+	return $access_token;
+}
+
+=head2 get_token_using_client_credentials()
+
+Gets a token for the user.
```

Comment:
Is it a token for the user, or a token for the application (acting as a user)

Because in the request I don't see the user mentioned !

If it's so, I would add a note to explain that this token must remain private and only be used server side for specific operations (like creating a new user).

Reply:
Generally, there's probably no reason to expose any tokens that exist in the server's RAM. Even the tokens that I'm currently logging for development purposes should be redacted later, so that they're not exposed in logs.

This specific token uses the server's client secret, and should definitely be kept private, because it will have admin-like permissions to create/edit users in keycloak, because of the `create_user_in_keycloak` sub.